README.md
Decred Cookbook
===============
[![Cookbook](https://img.shields.io/cookbook/v/decred.svg)](https://supermarket.getchef.com/cookbooks/decred)
[![Build Status](https://travis-ci.org/infertux/chef-decred.svg?branch=master)](https://travis-ci.org/infertux/chef-decred)

This cookbook installs and configures various Decred software.

Requirements
------------

### network
In order to actively contribute to the Decred network, you will need to open your TCP port 9108.
This cookbook does *not* make sure your port 9108 is open since this is very much dependant on your networking setup.

Usage
-----

### `decred::dcrd` recipe

Install and configure the [Decred daemon](https://github.com/decred/dcrd).

### `decred::dcrwallet` recipe

Install and configure the [Decred CLI wallet](https://github.com/decred/dcrwallet).

Development
-----------

```
bundle install
bundle exec berks install
bundle exec rake
```

In order to use https://github.com/test-kitchen/kitchen-digitalocean, you will need to set two env variables:

- DIGITALOCEAN_ACCESS_TOKEN
- DIGITALOCEAN_SSH_KEY_IDS

The Rakefile will automatically spin up DigitalOcean instances to run Inspec tests if the variables are defined.

License
-------
MIT


CHANGELOG.md
CHANGELOG
=========

## v1.0.0

- Initial release


test.sh
#!/bin/bash -eux

bundle update
bundle exec berks install
bundle exec berks update
bundle exec kitchen converge
bundle exec kitchen verify
# bundle exec kitchen destroy


test/integration/default/inspec/dcrd_spec.rb
control 'dcrd-1' do
  title 'dcrd'
  impact 1.0

  describe command('sudo -iu decred -- dcrd --version') do
    its('exit_status') { should eq 0 }
    its('stdout') { should match(/\Adcrd version 1\.1\.0\+release/) }
  end

  describe processes('dcrd') do
    its('entries.length') { should eq 1 }
    its('users') { should eq ['decred'] }
  end

  describe file('/home/decred/.dcrd/logs/mainnet/dcrd.log') do
    it { should be_file }
  end

  describe port(9108) do
    it { should_not be_listening }
  end

  describe port(9109) do
    it { should be_listening }
    its('protocols') { should cmp 'tcp' }
    its('addresses') { should cmp %w[127.0.0.1] }
  end
end


test/integration/default/inspec/dcrwallet_spec.rb
control 'dcrwallet-1' do
  title 'dcrwallet'
  impact 1.0

  describe processes('dcrwallet') do
    its('entries.length') { should eq 1 }
    its('users') { should eq ['decred'] }
  end

  describe file('/home/decred/.dcrwallet/logs/mainnet/dcrwallet.log') do
    it { should be_file }
  end

  describe port(9110) do
    it { should be_listening }
    its('protocols') { should cmp 'tcp' }
    its('addresses') { should cmp %w[127.0.0.1] }
  end

  describe port(9111) do
    it { should be_listening }
    its('protocols') { should cmp 'tcp' }
    its('addresses') { should cmp %w[127.0.0.1] }
  end
end


recipes/golang.rb
include_recipe 'golang::default'
include_recipe 'decred::_common'

%w[/ /bin /src].each do |suffix|
  directory "#{node['decred']['gopath']}#{suffix}" do
    owner node['decred']['user']
    group node['decred']['group']
  end
end

file "#{node['decred']['home']}/.profile" do
  owner node['decred']['user']
  group node['decred']['group']
  mode '0644'
end

ruby_block 'export GOPATH and GOBIN' do
  block do
    file = Chef::Util::FileEdit.new("#{node['decred']['home']}/.profile")

    [
      'export GOPATH=~/go',
      'export GOBIN=~/go/bin',
      'export PATH=$PATH:~/go/bin'
    ].each do |line|
      file.insert_line_if_no_match(Regexp.escape(line), line)
    end

    file.write_file
  end
end

bash 'check Go version' do
  flags '-l'
  user node['decred']['user']
  group node['decred']['group']
  environment('HOME' => node['decred']['home'])
  code <<-BASH
    set -eux
    go version
    go version | grep -Eq '^go version go#{node['go']['version']}(\.[0-9])? '
  BASH
end


recipes/_common.rb
# DigitalOcean apt caches are out-of-date # TODO: remove
apt_update 'update' do
  action :update
end

ohai 'reload_passwd' do
  action :nothing
  plugin 'etc'
end

group node['decred']['group']

user node['decred']['user'] do
  manage_home true
  group node['decred']['group']
  home node['decred']['home']
  shell '/bin/bash'
  notifies :reload, 'ohai[reload_passwd]', :immediately
end

directory "#{node['decred']['home']}/.ssh" do
  owner node['decred']['user']
  group node['decred']['group']
  mode '0700'
end

file "#{node['decred']['home']}/.ssh/authorized_keys" do
  owner node['decred']['user']
  group node['decred']['group']
  mode '0400'
  content node['decred']['ssh_authorized_keys']
end

file '/etc/logrotate.d/decred' do
  owner 'root'
  group 'root'
  mode '0444'
  content <<-CONF.gsub(/^\s+/, '')
    #{node['decred']['home']}/.*/logs/*/*.log {
      weekly
      rotate 52
      compress
      delaycompress
    }
  CONF
end


recipes/dcrinstall.rb
include_recipe 'decred::_common'
include_recipe 'decred::golang'

remote_file "#{node['decred']['home']}/dcrinstall" do
  owner node['decred']['user']
  group node['decred']['user']
  mode '0755'
  source "https://github.com/decred/decred-release/releases/download/v#{node['decred']['dcrinstall']['version']}/dcrinstall-linux-amd64-v#{node['decred']['dcrinstall']['version']}"
  checksum node['decred']['dcrinstall']['checksum']
  notifies :run, 'bash[dcrinstall]', :immediately
end

bash 'dcrinstall' do
  action :nothing
  user node['decred']['user']
  group node['decred']['user']
  environment('HOME' => node['decred']['home'])
  cwd node['decred']['home']
  code <<-BASH
    set -eux
    mkdir -p .dcrwallet/#{node['decred']['network']}
    touch .dcrwallet/#{node['decred']['network']}/wallet.db # workaround to skip wallet creation
    # XXX: https://github.com/decred/decred-release/blob/15d1f608e21aa1afbd05689a2454dbe49af294e8/cmd/dcrinstall/dcrinstall.go#L612
    ./dcrinstall
    grep decred .profile || echo 'export PATH=$PATH:~/decred' >> .profile
    test -s .dcrwallet/#{node['decred']['network']}/wallet.db || rm -rfv .dcrwallet/#{node['decred']['network']}
  BASH
end


recipes/dcrwallet.rb
include_recipe 'decred::_common'
include_recipe 'decred::dcrinstall'
include_recipe 'decred::dcrd'

template "#{node['decred']['home']}/.dcrwallet/dcrwallet.conf" do
  owner node['decred']['user']
  group node['decred']['group']
  mode '0400'
  notifies :restart, 'systemd_unit[dcrwallet.service]'
end

template "#{node['decred']['home']}/.dcrctl/dcrctl.conf" do
  owner node['decred']['user']
  group node['decred']['group']
  mode '0400'
end

directory "#{node['decred']['home']}/.dcrwallet/#{node['decred']['network']}" do
  owner node['decred']['user']
  group node['decred']['group']
  mode '0700'
end

file "#{node['decred']['home']}/.dcrwallet/passphrase" do
  owner node['decred']['user']
  group node['decred']['group']
  mode '0400'
  content node['decred']['dcrwallet']['passphrase']
  sensitive true
end

cookbook_file "#{node['decred']['home']}/.dcrwallet/#{node['decred']['network']}/wallet.db" do
  source 'dummywallet.db'
  owner node['decred']['user']
  group node['decred']['group']
  mode '0600'
  only_if { node['decred']['dcrwallet']['dummywallet'] }
end

systemd_unit 'dcrwallet.service' do
  content <<-SYSTEMD.gsub(/^\s+/, '')
    [Unit]
    Description=Decred dcrwallet
    ConditionPathExists=#{node['decred']['home']}/.dcrwallet/#{node['decred']['network']}/wallet.db

    [Service]
    Type=simple
    User=#{node['decred']['user']}
    Group=#{node['decred']['group']}
    WorkingDirectory=#{node['decred']['home']}
    ExecStart=/bin/bash -c 'exec #{node['decred']['home']}/decred/dcrwallet < #{node['decred']['home']}/.dcrwallet/passphrase'
    Restart=on-failure

    [Install]
    WantedBy=multi-user.target
  SYSTEMD

  action %i[create enable start]
end

log 'check for wallet.db' do
  level :warn
  message "No wallet.db found - log in via ssh then run `dcrwallet --create` as the #{node['decred']['user']} user."
  not_if { File.exist?("#{node['decred']['home']}/.dcrwallet/#{node['decred']['network']}/wallet.db") }
end
